File: sdks/python3/blok/middleware/recovery_middleware.py
from __future__ import annotations
import traceback
from typing import Any, Callable, Dict

from blok.errors.blok_error import BlokError
from blok.errors.node_error import NodeError, ErrorCategory


def recovery_middleware(next_fn: Callable) -> Callable:
    """Middleware that catches unhandled exceptions and converts them to NodeErrors.

    ``BlokError`` instances pass through untouched — they're already
    structured per master plan §17 and would lose every field except
    ``message`` if wrapped as a legacy ``NodeError``. Same with the legacy
    ``NodeError``: pre-wrapped errors keep their shape.
    """

    def wrapper(ctx, config: Dict[str, Any]) -> Any:
        try:
            return next_fn(ctx, config)
        except BlokError:
            # Structured BlokError: pass through verbatim so the registry
            # + gRPC servicer can serialize every field losslessly.
            raise
        except NodeError:
            raise
        except Exception as e:
            raise NodeError(
                message=f"unhandled exception: {e}",
                code=500,
                category=ErrorCategory.EXECUTION,
                details={"traceback": traceback.format_exc()},
                cause=e,
            ) from e

    return wrapper

File: sdks/ruby/lib/blok/middleware/recovery_middleware.rb
# frozen_string_literal: true

module Blok
  module Middleware
    # RecoveryMiddleware rescues unhandled exceptions and converts them
    # into structured error results, preventing the runtime from crashing.
    class RecoveryMiddleware < Middleware
      # Wrap the handler with exception recovery.
      # @param handler [#call] Inner handler callable
      # @return [#call] Wrapped callable with recovery
      def wrap(handler)
        ->(ctx, config) {
          begin
            handler.call(ctx, config)
          rescue Errors::BlokError
            # Master plan §17 BlokError passes through verbatim — the
            # registry catches it directly and stashes the typed instance
            # on `ExecutionResult.errors` for the gRPC servicer.
            raise
          rescue Errors::NodeError
            # Legacy structured exceptions pass through as-is too.
            raise
          rescue StandardError => e
            # Convert unexpected errors into a structured error response
            raise Errors::NodeError.execution("unexpected error: #{e.message}")
          end
        }
      end
    end
  end
end
